auto-value-gson/src/test/java/com/ryanharter/auto/value/gson/AutoValueGsonExtensionTest.java

```diff
@@ -809,4 +812,121 @@ public void simple() {
+        + "    private String defaultB = null;\n"
+        + "    private String defaultC = null;\n"
+        + "    private String defaultD = null;\n"
+        + "    public TestTypeAdapter(Gson gson) {\n"
+        + "      this.aAdapter = gson.getAdapter(String.class);\n"
+        + "      this.bAdapter = gson.getAdapter(String.class);\n"
+        + "      this.cAdapter = gson.getAdapter(String.class);\n"
+        + "    }\n"
+        + "    public GsonTypeAdapter setDefaultA(String defaultA) {\n"
+        + "      this.defaultA = defaultA;\n"
+        + "      return this;\n"
+        + "    }\n"
+        + "    public GsonTypeAdapter setDefaultB(String defaultB) {\n"
+        + "      this.defaultB = defaultB;\n"
+        + "      return this;\n"
+        + "    }\n"
+        + "    public GsonTypeAdapter setDefaultC(String defaultC) {\n"
+        + "      this.defaultC = defaultC;\n"
+        + "      return this;\n"
+        + "    }\n"
+        + "    public GsonTypeAdapter setDefaultD(String defaultD) {\n"
+        + "      this.defaultD = defaultD;\n"
+        + "      return this;\n"
+        + "    }\n"
+        + "    @Override\n"
+        + "    public void write(JsonWriter jsonWriter, Test object) throws IOException {\n"
+        + "      if (object == null) {\n"
+        + "        jsonWriter.nullValue();\n"
+        + "        return;\n"
+        + "      }\n"
+        + "      jsonWriter.beginObject();\n"
+        + "      jsonWriter.name(\"a\");\n"
+        + "      aAdapter.write(jsonWriter, object.a());\n"
+        + "      jsonWriter.name(\"b\");\n"
+        + "      bAdapter.write(jsonWriter, object.b());\n"
+        + "      jsonWriter.endObject();\n"
+        + "    }\n"
+        + "    @Override\n"
+        + "    public Test read(JsonReader jsonReader) throws IOException {\n"
+        + "      if (jsonReader.peek() == JsonToken.NULL) {\n"
+        + "        jsonReader.nextNull();\n"
+        + "        return null;\n"
+        + "      }\n"
+        + "      jsonReader.beginObject();\n"
+        + "      String a = this.defaultA;\n"
+        + "      String b = this.defaultB;\n"
+        + "      String c = this.defaultC;\n"
+        + "      String d = this.defaultD;\n"
+        + "      while (jsonReader.hasNext()) {\n"
+        + "        String _name = jsonReader.nextName();\n"
+        + "        switch (_name) {\n"
+        + "          case \"a\": {\n"
+        + "            a = aAdapter.read(jsonReader);\n"
+        + "            break;\n"
+        + "          }\n"
+        + "          case \"c\": {\n"
+        + "            c = cAdapter.read(jsonReader);\n"
+        + "            break;\n"
+        + "          }\n"
+        + "          default: {\n"
+        + "            jsonReader.skipValue();\n"
+        + "          }\n"
+        + "        }\n"
+        + "      }\n"
+        + "      jsonReader.endObject();\n"
+        + "      return new AutoValue_Test(a, b, c, d);\n"
```

Comment:
I don't think that's true. First, it won't fail to compile, since AutoValue allows you to pass null (it will throw in runtime though). Second, the value of `d` here is not always null, it's actually whatever you've set as default for that field with `setDefaultD`. We obviously can't tell if the default value has been set or not during codegen, so I don't think there's a good way to handle that here.

Basically, the effect is that if you mark something as ignored for deserialization *and* you haven't annotated it `@Nullable` *and* you haven't set a non-null default value for said field on the generated adapter, deserialization will fail in runtime with an NPE. To me, this is expected behavior.

The alternative is to remove the concept of default values specifically for fields that are ignored by deserialization, and enforce them to deserialize as null - then we can enforce that said property is marked as nullable during codegen. This seems unnecessarily restrictive though, IMO.

If I misunderstood something, let me know :)